File: submissions/0525-contiguous-array/solution.ts
function findMaxLength(nums: number[]): number {
    let maxLength = 0;
    let count = 0;
    const countMap = new Map<number, number>();

    countMap.set(0, -1);

    for (let i = 0; i < nums.length; i++) {
        // Increment or decrement count
        count += nums[i] === 1 ? 1 : -1;

        // Check if this count has been seen before
        if (countMap.has(count)) {
            // Calculate the length of the subarray
            const prevIndex = countMap.get(count)!;
            maxLength = Math.max(maxLength, i - prevIndex);
        } else {
            // Store the first occurrence of this count with its index
            countMap.set(count, i);
        }
    }

    return maxLength;
}

File: submissions/0525-contiguous-array/solution.py
from typing import List

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        balance_index_map = {0: -1}  
        max_length = 0
        balance = 0
        
        for i, num in enumerate(nums):
            balance += 1 if num == 1 else -1
            
            if balance in balance_index_map:
                max_length = max(max_length, i - balance_index_map[balance])
            else:
                balance_index_map[balance] = i
        
        return max_length

